File: problems/leetcode/0300/377.cpp
// 377. Combination Sum IV
// https://leetcode.com/problems/combination-sum-iv/description/

using namespace std;

#include <unordered_map>
#include <vector>

class Solution {
public:
  unordered_map<int, int> memo;

  int combinationSum4(vector<int> &nums, int target) {
    if (target == 0) {
      return 1;
    }

    if (memo.find(target) != memo.end()) {
      return memo[target];
    }

    memo[target] = 0;
    for (int num : nums) {
      if (target - num >= 0) {
        memo[target] += combinationSum4(nums, target - num);
      }
    }

    return memo[target];
  }
};

File: problems/leetcode/0300/377/377.py
# 377. Combination Sum IV
# https://leetcode.com/problems/combination-sum-iv/

class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = dict()

        def getCombinationSum(t: int) -> int:
            if t in dp:
                return dp[t]

            dp[t] = 0

            for num in nums:
                if num == t:
                    dp[t] += 1
                elif num < t:
                    dp[t] += getCombinationSum(t - num)

            return dp[t]

        return getCombinationSum(target)
